core/app/models/spree/address.rb

```diff
@@ -18,19 +18,48 @@ class Address < Spree::Base
     alias_attribute :first_name, :firstname
     alias_attribute :last_name, :lastname
 
+    DB_ONLY_ATTRS = %w(id updated_at created_at)
+
+    scope :with_values, ->(attributes) do
+      where(value_attributes(attributes))
+    end
+
     def self.build_default
-      country = Spree::Country.find(Spree::Config[:default_country_id]) rescue Spree::Country.first
-      new(country: country)
+      new(country: Spree::Country.default)
     end
 
     def self.default(user = nil, kind = "bill")
-      if user && user_address = user.send(:"#{kind}_address")
-        user_address.dup
+      ActiveSupport::Deprecation.warn("Address.default is deprecated. Use User.default_address or Address.build_default", caller)
+      if user
+        user.send(:"#{kind}_address") || build_default
       else
         build_default
       end
     end
 
+    # @return [Address] address from existing address plus new_attributes as diff
+    # @note, this may return existing_address if there are no changes to value equality
+    def self.immutable_merge(existing_address, new_attributes)
+      return self.new(value_attributes(new_attributes)) unless existing_address
+
+      new_attributes ||= {}
+      merged_attributes = existing_address.attributes.merge(new_attributes.stringify_keys).except(*DB_ONLY_ATTRS)
+      new_address = self.new(merged_attributes)
+      if existing_address == new_address
+        existing_address
+      else
+        new_address
+      end
+    end
+
+    def self.value_attributes(attributes_hash)
+      (attributes_hash || {}).stringify_keys.except(*DB_ONLY_ATTRS)
```

Comment:
I'm not sure what you're getting at. Do you think we should create an Address instance here in order to validate and then return attributes? In one of the uses of this method (via `immutable_merge` -- line 42 above) the incoming Hash could be a list of only fields that should be changed, not a full set of fields to make up a valid Address itself.

This method isn't about validation. It's about taking a Hash which potentially may have extra fields and returning a list that we know is assignable to Address without ultimately causing unknown column errors when saved.